### PR TITLE
feat: check-in `cargo dylint` with `format_error` lint

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Check-in `cargo dylint` lints. See [`lints/README.md`](https://github.com/risingwavelabs/risingwave/pull/13750/files#diff-5503594d102c30b00449ac60ee89580c5ea23672aef2b9f59c1838fc62feac86) for more information.

As described in https://github.com/risingwavelabs/risingwave/pull/13725, it's not a good practice to directly format an error with `Display` or `Debug`. However, it could be quite painful for developers to manually find and correct them.

In this PR, we leverage `cargo dylint` to write a custom lint called `format_error` and check the usages. The experience for both development and check is quite similar to `clippy`, so it can be easily continuously integrated.

image
image
image

To keep the PR clean, I haven't integrated it to CI or RiseDev now. Will do this in next PRs.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: lints/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+[lib]
+crate-type = ["cdylib"]
+
+[[example]]
+name = "format_error"
+path = "ui/format_error.rs"
+
+[dependencies]
+clippy_utils = { git = "https://github.com/rust-lang/rust-clippy", rev = "a585cda701581a16894858dc088eacd5a02fc78b" } # should match the toolchain version
```
[the remark this answers is not in the document]
Comment (xxchan) [reply lints/Cargo.toml:15]: BTW, why does it need to match

**Review by xxchan — COMMENTED** [review]
I wanted to rubber stamp, but one thing cause me hesitant: `clippy` is already very slow, and `dylint` needs to rebuild the project once again. How to resolve this issue?

**Comment by liurenjie1024** [comment]
> I wanted to rubber stamp, but one thing makes me hesitant: `clippy` is already very slow, and `dylint` needs to rebuild the project once again. How to resolve this issue?

+1. Is it really worth it?

**Comment by TennyZhuang** [comment]
Perhaps one lint is not worth it, but I can think of many rules, so I weakly support this.

**Comment by TennyZhuang** [comment]
> and dylint needs to rebuild the project once again. How to resolve this issue

parallelize them?

**Comment by BugenZhao** [comment]
> I wanted to rubber stamp, but one thing makes me hesitant: `clippy` is already very slow, and `dylint` needs to rebuild the project once again. How to resolve this issue?

> +1. Is it really worth it?

It's not a question of worth or not because I can't think of any other way to prevent the loss of error information. 😕 

In theory, the slowness of `clippy` is unrelated to the performance of our lints. Note that `clippy` has hundreds of lints, while we only have one at present. Thus, the process only involves running `cargo check` and this single lint. If `check` can be cached, then it would be fast. I'll do further local testing.

If the running time is really a problem, I guess linting in main-cron or even manually is still good.

File: lints/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply lints/Cargo.toml:15]: The internal API of rustc is not stable, which `rust-clippy` depends on.

File: lints/src/utils/format_args_collector.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#[derive(Default)]
+pub struct FormatArgsCollector {
+    format_args: FxHashMap<Span, Rc<FormatArgs>>,
+}
+
+impl_lint_pass!(FormatArgsCollector => []);
+
+impl EarlyLintPass for FormatArgsCollector {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::FormatArgs(args) = &expr.kind {
```
Comment (stdrc) [inline lints/src/utils/format_args_collector.rs:44]: Is `format_args` a special macro which the compiler is aware of?

File: lints/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply lints/Cargo.toml:15]: > Maybe we will need to add a script or checklist comments in `rust-toolchain` for updating toolchain now 🤣

It's okay. Mismatched version of `clippy_utils` usually won't compile.

**Review by stdrc — APPROVED** [review]
The concept looks really good to me. Previously we have some times when we wanted to add some custom lint rules but gave up. Such a dylint integration will make future custom linting easy. Thx!

File: lints/rust-toolchain [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[toolchain]
+channel = "nightly-2023-10-21"                      # should be identical to the root one
```
Comment (liurenjie1024) [inline lints/rust-toolchain:2]: Is there anyway to sync this with root one automatically?

**Review by liurenjie1024 — APPROVED** [review]
LGTM

File: lints/src/utils/format_args_collector.rs (hunk 2 above)
Comment (BugenZhao) [reply lints/src/utils/format_args_collector.rs:44]: I guess so, at the early pass.

**Comment by xxchan** [comment]
> Whether or not we should integrate dylintto the development flow is a
topic in next PRs. In any case, checking in lints to our repository enable
collaboration on improving the quality of RisingWave's codebase

LGTM. Agree to check in and run daily and manually.

Btw, by slowness I did mean cargo check is slow, and dylint doesn’t share
build with clippy is worse.

**Pull request merged**